# **IMPORTING THE NECESSARY DATASET **

In [7]:
## Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline


url='https://github.com/WalePhenomenon/climate_change/blob/master/fuel_ferc1.csv?raw=true'
df = pd.read_csv(url, error_bad_lines=False)

In [9]:
#Viewing the head top 5 datas
df.head(5)

record_id  ...  fuel_cost_per_mmbtu
0   f1_fuel_1994_12_1_0_7  ...                1.121
1  f1_fuel_1994_12_1_0_10  ...                1.120
2   f1_fuel_1994_12_2_0_1  ...                1.650
3   f1_fuel_1994_12_2_0_7  ...                1.970
4  f1_fuel_1994_12_2_0_10  ...                2.570

[5 rows x 11 columns]

In [10]:
#Viewing the last 5 datas
df.tail(5)

record_id  ...  fuel_cost_per_mmbtu
29518  f1_fuel_2018_12_12_0_13  ...                 9.03
29519   f1_fuel_2018_12_12_1_1  ...                 6.95
29520  f1_fuel_2018_12_12_1_10  ...                 8.99
29521  f1_fuel_2018_12_12_1_13  ...                 1.11
29522  f1_fuel_2018_12_12_1_14  ...                11.68

[5 rows x 11 columns]

#The shape and info of datasets

In [14]:
print("Dataset shape :",df.shape)

df.info()

Dataset shape : (29523, 11)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


# **Question 1**



If you’re given two lists:

A = [1,2,3,4,5,6], B = [13, 21, 34]

The task is to create a list with the elements of A and B in a single dimension with output:

A_B = [1,2,3,4,5,6,13, 21, 34]

Which of the following option is the best way to create this list?

In [2]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]

A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

# **Question 2**


How do you create an identity matrix in python?

In [4]:
#Importing Libraries
import numpy as np 


np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# **Question 3**



Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [15]:
df.head()

record_id  ...  fuel_cost_per_mmbtu
0   f1_fuel_1994_12_1_0_7  ...                1.121
1  f1_fuel_1994_12_1_0_10  ...                1.120
2   f1_fuel_1994_12_2_0_1  ...                1.650
3   f1_fuel_1994_12_2_0_7  ...                1.970
4  f1_fuel_1994_12_2_0_10  ...                2.570

[5 rows x 11 columns]

In [16]:
df.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64


The average for gas is lowest


# **Question 4**



What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [17]:
round(df.describe(),2)

utility_id_ferc1  ...  fuel_cost_per_mmbtu
count          29523.00  ...             29523.00
mean             118.60  ...                19.30
std               74.18  ...              2091.54
min                1.00  ...               -41.50
25%               55.00  ...                 1.94
50%              122.00  ...                 4.13
75%              176.00  ...                 7.74
max              514.00  ...            359278.00

[8 rows x 7 columns]

For fuel_mmbtu_per_unit the Std dev is 10.60 and 75th Percentile is 17.01.


# **Question 5**



What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [18]:
round(df.skew(),2)

utility_id_ferc1                  0.61
report_year                       0.01
fuel_qty_burned                  15.85
fuel_mmbtu_per_unit               4.14
fuel_cost_per_unit_burned        19.79
fuel_cost_per_unit_delivered    105.01
fuel_cost_per_mmbtu             171.68
dtype: float64

In [19]:
round(df.kurtosis(),2)


utility_id_ferc1                    1.09
report_year                        -1.15
fuel_qty_burned                   651.37
fuel_mmbtu_per_unit                55.60
fuel_cost_per_unit_burned         485.26
fuel_cost_per_unit_delivered    11765.05
fuel_cost_per_mmbtu             29489.13
dtype: float64

For Fuel_qty_burned Skewness is 15.85 and Kurtosis is 651.37

# **Question 6**




Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places?

(Enter answer in the format: Feature: xxx, Total: xxx, Percent: xxx)



In [20]:
feature = [col for col in df.columns if df[col].isnull().any()]
total = df['fuel_unit'].isnull().sum()
percent_missing = round(df['fuel_unit'].isnull().sum() * 100 / len(df),4)
print('Feature:',feature,', ''Total',total,', ''Percent',percent_missing)

Feature: ['fuel_unit'] , Total 180 , Percent 0.6097


# **Question 7**



The feature with missing values falls under what category? What missing value imputation technique would you use?


ANS:- The feature with missing values falls under Categorical category.
Missing value imputation technique used would be Mode imputation

# **Question 8**



Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [21]:
df.corr()['fuel_cost_per_unit_burned'].sort_values(ascending=True)

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

fuel_qty_burned and fuel_mmbtu_per_unit have second and third lowest correlation respectively with the Fuel Cost Per Unit Burned

# **Question 9**



For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [22]:
df.head()


record_id  ...  fuel_cost_per_mmbtu
0   f1_fuel_1994_12_1_0_7  ...                1.121
1  f1_fuel_1994_12_1_0_10  ...                1.120
2   f1_fuel_1994_12_2_0_1  ...                1.650
3   f1_fuel_1994_12_2_0_7  ...                1.970
4  f1_fuel_1994_12_2_0_10  ...                2.570

[5 rows x 11 columns]

In [23]:

a = df.loc[(df['report_year']==1998) & (df['fuel_type_code_pudl']=='coal')]['fuel_cost_per_unit_burned'].sum()
a

11902.597

In [24]:
b = df.loc[(df['report_year']==1994) & (df['fuel_type_code_pudl']=='coal')]['fuel_cost_per_unit_burned'].sum()
b

14984.572000000002

In [25]:
print('The percentage change from 1998 compared to 1994:', round((a-b)*100/a))

The percentage change from 1998 compared to 1994: -26


# **Question 10**



Which year has the highest average fuel cost per unit delivered?

In [28]:
df.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

Year 1997 has the highest average fuel cost per unit delivered.